File: scripts/validate-durababble-sigils.rb
#!/usr/bin/env ruby
# typed: true
# frozen_string_literal: true

# Validates that all [DURABABBLE-*] formal-model sigils appear in the Ruby
# implementation/specs and that Ruby-side sigils exist in the Alloy model.

require "find"
require "pathname"
require "set"

TAG_REGEX = /\[DURABABBLE-[A-Z0-9-]+\]/
MODEL_DIRS = ["formal"].freeze
RUBY_DIRS = ["lib", "test"].freeze

def find_root(start)
  dir = Pathname.new(start).expand_path

  until (dir + "durababble.gemspec").exist?
    parent = dir.parent
    raise "Could not find repository root (missing durababble.gemspec)" if parent == dir

    dir = parent
  end

  dir
end

def find_files(root, dirs)
  dirs.flat_map do |dir|
    full_dir = root + dir
    next [] unless full_dir.directory?

    files = []
    Find.find(full_dir.to_s) do |path|
      pathname = Pathname.new(path)
      files << pathname if pathname.file? && yield(pathname)
    end
    files
  end
end

def extract(files, root)
  tags = Set.new
  locations = Hash.new { |hash, key| hash[key] = [] }

  files.each do |file|
    file.read.scan(TAG_REGEX).each do |tag|
      tags.add(tag)
      locations[tag] << file.relative_path_from(root).to_s
    end
  end

  [tags, locations]
end

def difference(left, right)
  left.reject { |tag| right.include?(tag) }.sort
end

def print_tag_list(title, tags, locations)
  puts title
  tags.each do |tag|
    puts "  #{tag} (#{locations[tag].join(", ")})"
  end
end

def print_matched(tags, model_locations, ruby_locations)
  puts "Matched tags:"
  tags.each do |tag|
    puts "  #{tag}"
    puts "    Alloy: #{model_locations[tag].join(", ")}"
    puts "    Ruby:  #{ruby_locations[tag].join(", ")}"
  end
end

root = find_root(Dir.pwd)
model_files = find_files(root, MODEL_DIRS) { |file| file.extname == ".als" }
ruby_files = find_files(root, RUBY_DIRS) { |file| file.extname == ".rb" }

if model_files.empty?
  warn "No Alloy model files found under formal/"
  exit 1
end

if ruby_files.empty?
  warn "No Ruby implementation/spec files found under lib/ or test/"
  exit 1
end

model_tags, model_locations = extract(model_files, root)
ruby_tags, ruby_locations = extract(ruby_files, root)
only_in_model = difference(model_tags, ruby_tags)
only_in_ruby = difference(ruby_tags, model_tags)
in_both = model_tags.select { |tag| ruby_tags.include?(tag) }.sort

puts "Durababble sigil validation"
puts "============================"
puts "Alloy files: #{model_files.length}"
puts "Ruby files:  #{ruby_files.length}"
puts "Matched tags: #{in_both.length}"
puts "Only in Alloy: #{only_in_model.length}"
puts "Only in Ruby:  #{only_in_ruby.length}"
puts

failed = false
unless only_in_model.empty?
  failed = true
  print_tag_list("Tags in Alloy but missing from Ruby:", only_in_model, model_locations)
  puts
end

unless only_in_ruby.empty?
  failed = true
  print_tag_list("Tags in Ruby but missing from Alloy:", only_in_ruby, ruby_locations)
  puts
end

print_matched(in_both, model_locations, ruby_locations) if ARGV.include?("--verbose") || ARGV.include?("-v")

exit(failed ? 1 : 0)

File: scripts/validate-durababble-sigils.js
#!/usr/bin/env node
/*
 * Validates that all [DURABABBLE-*] formal-model sigils appear in the Ruby
 * implementation/specs and that Ruby-side sigils exist in the Alloy model.
 */

const fs = require("fs");
const path = require("path");

const TAG_REGEX = /\[DURABABBLE-[A-Z0-9-]+\]/g;
const MODEL_DIRS = ["formal"];
const RUBY_DIRS = ["lib", "test"];

function findRoot(start) {
  let dir = start;
  while (!fs.existsSync(path.join(dir, "durababble.gemspec"))) {
    const parent = path.dirname(dir);
    if (parent === dir) {
      throw new Error("Could not find repository root (missing durababble.gemspec)");
    }
    dir = parent;
  }
  return dir;
}

function findFiles(root, dirs, predicate, results = []) {
  for (const dir of dirs) {
    const fullDir = path.join(root, dir);
    if (!fs.existsSync(fullDir)) continue;
    walk(fullDir, predicate, results);
  }
  return results;
}

function walk(dir, predicate, results) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walk(fullPath, predicate, results);
    } else if (entry.isFile() && predicate(fullPath)) {
      results.push(fullPath);
    }
  }
}

function extract(files, root) {
  const tags = new Set();
  const locations = new Map();

  for (const file of files) {
    const text = fs.readFileSync(file, "utf8");
    const matches = text.match(TAG_REGEX) || [];
    for (const tag of matches) {
      tags.add(tag);
      if (!locations.has(tag)) locations.set(tag, []);
      locations.get(tag).push(path.relative(root, file));
    }
  }

  return { tags, locations };
}

function difference(left, right) {
  return [...left].filter((tag) => !right.has(tag)).sort();
}

function printTagList(title, tags, locations) {
  console.log(title);
  for (const tag of tags) {
    console.log(`  ${tag} (${(locations.get(tag) || []).join(", ")})`);
  }
}

function printMatched(tags, modelLocations, rubyLocations) {
  console.log("Matched tags:");
  for (const tag of tags) {
    console.log(`  ${tag}`);
    console.log(`    Alloy: ${(modelLocations.get(tag) || []).join(", ")}`);
    console.log(`    Ruby:  ${(rubyLocations.get(tag) || []).join(", ")}`);
  }
}

function main() {
  const root = findRoot(process.cwd());
  const modelFiles = findFiles(root, MODEL_DIRS, (file) => file.endsWith(".als"));
  const rubyFiles = findFiles(root, RUBY_DIRS, (file) => file.endsWith(".rb"));

  if (modelFiles.length === 0) {
    console.error("No Alloy model files found under formal/");
    process.exit(1);
  }
  if (rubyFiles.length === 0) {
    console.error("No Ruby implementation/spec files found under lib/ or test/");
    process.exit(1);
  }

  const model = extract(modelFiles, root);
  const ruby = extract(rubyFiles, root);
  const onlyInModel = difference(model.tags, ruby.tags);
  const onlyInRuby = difference(ruby.tags, model.tags);
  const inBoth = [...model.tags].filter((tag) => ruby.tags.has(tag)).sort();

  console.log("Durababble sigil validation");
  console.log("============================");
  console.log(`Alloy files: ${modelFiles.length}`);
  console.log(`Ruby files:  ${rubyFiles.length}`);
  console.log(`Matched tags: ${inBoth.length}`);
  console.log(`Only in Alloy: ${onlyInModel.length}`);
  console.log(`Only in Ruby:  ${onlyInRuby.length}`);
  console.log();

  let failed = false;
  if (onlyInModel.length > 0) {
    failed = true;
    printTagList("Tags in Alloy but missing from Ruby:", onlyInModel, model.locations);
    console.log();
  }

  if (onlyInRuby.length > 0) {
    failed = true;
    printTagList("Tags in Ruby but missing from Alloy:", onlyInRuby, ruby.locations);
    console.log();
  }

  if (process.argv.includes("--verbose") || process.argv.includes("-v")) {
    printMatched(inBoth, model.locations, ruby.locations);
  }

  process.exit(failed ? 1 : 0);
}

main();
